lib/cog/command/pipeline/destination.ex

```diff
@@ -0,0 +1,123 @@
+    else
+      [%{make_destination("here") | output_level: :status_only} | destinations]
+    end
+  end
+
+  defp originator_is_destination?(destinations),
+    do: Enum.any?(destinations, &(&1.raw == "here"))
+
+  # Resolve all destinations according to their specified adapter. If
+  # all are valid, returns an `{:ok, destinations}` tuple containing
+  # all fully-resolved destinations. If any fail to resolve, an
+  # `{:error, reasons}` tuple is returned, containing only the errors
+  # for the failing destinations.
+  defp resolve(destinations, sender, origin_room, origin_adapter) do
+    Enum.reduce(destinations, {:ok, []}, fn(destination, {last_status, values}=acc) ->
+      {current_status, value} = resolve_destination(destination, sender, origin_room, origin_adapter)
+      case {last_status, current_status} do
+        {:ok, :ok} -> # Everything is good; collect
+          {:ok, [value|values]}
+        {:ok, :error} -> # We hit the first error; discard everything before
+          {:error, [value]}
+        {:error, :error} -> # Still hitting errors; collect
+          {:error, [value|values]}
+        {:error, :ok} -> # Ignore successful resolutions when we've encountered errors before
+          acc
+      end
+    end)
+  end
+
+  defp resolve_destination(%__MODULE__{raw: "here"}=dest, _sender, origin_room, adapter),
+    do: {:ok, %{dest | adapter: adapter.name, room: origin_room}}
+  defp resolve_destination(%__MODULE__{raw: "me"}=dest, sender, _origin_room, adapter) do
+    user_id = sender["id"]
+    case adapter.lookup_direct_room(user_id: user_id) do # TODO: user should be opaque to all but adapter?
+      {:ok, direct_chat} ->
+        {:ok, %{dest | adapter: adapter.name, room: direct_chat}}
+      error ->
+        Logger.error("Error resolving redirect 'me' with adapter #{adapter}: #{inspect error}")
+        {:error, "me"}
+    end
+  end
+  defp resolve_destination(%__MODULE__{raw: redir}=dest, _sender, _origin_room, origin_adapter) do
+    {adapter, destination} = adapter_destination(redir, origin_adapter)
+    case adapter.lookup_room(destination) do
+      {:ok, room} ->
+        if adapter.room_writeable?(id: room.id) == true do # TODO: just pass the room entirely
+          {:ok, %{dest | adapter: adapter.name, room: room}}
+        else
+          {:error, {:not_a_member, redir}}
+        end
+      {:error, reason} ->
+        Logger.error("Error resolving redirect '#{redir}' with adapter #{adapter}: #{inspect reason}")
+        {:error, {reason, redir}}
+    end
+  end
+
+  # Redirect destinations may be targeted to an adapter different from
+  # where they originated from.
+  #
+  # Destinations prefixed with "chat://" will be routed through the
+  # active chat adapter module. Anything else will be routed through
+  # the adapter that initially serviced the request.
+  @spec adapter_destination(String.t, atom) :: {atom, String.t}
+  defp adapter_destination("chat://" <> destination, _origin_adapter) do
```

Comment:
Is there a potential to have multiple active chat adapters? If so, would we want to route based on something like `slack://` or `hipchat://`?